File: 最大公约数和最小公倍数.py
## 辗转相除法

# 计算 gcd(18,48) 时，先将 48 除以 18 得到商 2、余数 12，然后再将 18 除以 12 得到商 1、余数 6，再将 12 除以 6 得到商 2、余数 0，即得到最大公因数 6 。我们只关心每次除法的余数是否为 0，为 0 即表示得到答案。
# 原理：如果
# rem = x%y
# 如果 div 是 y 和 x%y 的最大公约数
# 那 div 一定也是 x 的公约数，因为 x = x % y + m * y
# div 能被 x%y 和 y 整除，x 就一定可以

def gcd(x, y):
    if y == 0:
        return x
    if x < y:
        x, y = y, x
    return gcd(y, x%y)

t1 = 10
t2 = 15

lcm = t1 * t2 / gcd(t1, t2)

print("least common multiple of {} and {} is\n {}".format(t1, t2, lcm))


# pythonic 但是结果不对，不满足 x > y

def gcd(x, y):
    while y:
        x, y = y, x%y
    return x

File: 最大公约数和最小公倍数.cpp
// 两个数的最小公倍数(Lowest Common Multiple) = 两个数的乘积 / 两个数的最大公约数

// 在数论中，如果两个或两个以上的整数的最大公约数是 1，则称它们为互质
// 最大公约数（英语：Greatest Common Divisor，简写为G.C.D.；或Highest Common Factor，简写为H.C.F.）


// 1不是质数

// 求最大公约数：欧几里得法（辗转相除法）
// 求a和b的最大公约数m
// 我们知道 a = kb + r，可以表示任意两个整数之间的关系，且 a%m == b%m == 0，所以 r%m == 0
// 也就是说 m 也是 b 和 r 的最大公约数

#include <iostream>
using namespace std;

int Gcd(int x, int y)
{
    if (y == 0)
        return x;
    else
        Gcd(y, x%y);
}



int main(int argc, char const *argv[]) {
    int a, b;
    cin >> a >> b;

    int gcd = Gcd(a, b);
    int lcm = a * b / gcd;

    cout << "gcd: " << gcd << endl;
    cout << "lcm: " << lcm << endl;
    
    return 0;
}
